Fix deletes not reaching every server that interacted with status

Gargron opened the pull request:
Extract logic for determining ActivityPub inboxes to send deletes to to its own class and explicitly include the person the status
replied to (even if not mentioned), people who favourited it, and people who replied to it (though that one is still not recursive)

Gargron committed:
Fix deletes not reaching every server that interacted with status

Extract logic for determining ActivityPub inboxes to send deletes
to to its own class and explicitly include the person the status
replied to (even if not mentioned), people who favourited it, and
people who replied to it (though that one is still not recursive)

ThibG commented:
I'm ok with the general idea, and factoring it out looks cleaner and is easier to reason about. As for the failing test, I have to admit I have no clue what's going on, for some reason I don't understand, `@status.reblogs` is empty even though a reblog has been instantiated.

It should also be noted that instances that fetched the toot in any way without actually interacting with it won't necessarily get a delete, and it mostly relies on intermediary instances forwarding the `Delete`. But it should be an improvement over the status quo nonetheless.

Gargron commented:
The unit test caught a real issue. Reblogs were removed before `StatusReachFinder` was called, this was fine before because reblogs were preloaded into an instance variable. But I see no reason why the order of execution should be that way anymore, so I changed it. Also, since the preloads were no longer used more than once in the class, I replaced them with potentially more efficient `find_each` calls.

Gargron commented:
I also identified some inefficies. When removing a reblog, the class would make queries for the reblog's mentions, media attachments, tags, etc. Since reblogs don't possess those, we can cut down on those queries. Also, due to using `pluck` on the tags, the tags actually got queried twice, once the IDs for featured tags, and then the names for the timelines.

ThibG commented on app/services/remove_status_service.rb:
```diff
@@ -67,31 +70,35 @@ def remove_from_lists
+    # followers. Here we send a delete to actively mentioned accounts
+    # that may not follow the account
+
+    @status.active_mentions.find_each do |mention|
+      redis.publish("timeline:#{mention.account_id}", @payload)
     end
```
I understand this is to avoid fetching account information from the database, but this will send events about remote accounts to redis. Is it worth it? Does it make sense?

Gargron replied:
You’re right... but publishing to a channel with no subscribers should be O(0) so probably fine this way.

ClearlyClaire reviewed (approved):
Good catch with the reblogs stuff. This seems better overall, but there's one thing I'm not sure about (though it shouldn't be a big issue), see inline comments.
